### PR TITLE
Tcp mt

### PR DESCRIPTION
**Pull request by ooststep**

### CODE REVIEW

File: prov/util/src/util_av.c [hunk 1]
```diff
@@ -313,6 +341,7 @@ int ofi_av_insert_addr(struct util_av *av, const void *addr, fi_addr_t *fi_addr)
 		FI_INFO(av->prov, FI_LOG_AV, "fi_addr: %" PRIu64 "\n",
 			ofi_buf_index(entry));
 	}
+
```
Comment (aingerson) [inline prov/util/src/util_av.c]: drop line

File: prov/tcp/src/xnet.h [hunk 2]
```diff
@@ -467,12 +467,67 @@ struct xnet_xfer_entry {
...
+}
+
+int xnet_domain_multiplexed(struct fid_domain *domain_fid);
+
+int xnet_multiplex_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
+			   struct fid_av **fid_av, void *context);
+
+int xnet_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		     struct fid_domain **domain, void *context);
+
```
Comment (aingerson) [inline prov/tcp/src/xnet.h]: Duplicate definitions of these two calls. I also think the only reason xnet_domain_open was moved up here was to call in create_subdomain which I think can be removed so likely need to undo this change

File: prov/tcp/src/xnet.h [hunk 3]
```diff
@@ -467,12 +467,67 @@ struct xnet_xfer_entry {
...
 };
 
+int xnet_multiplex_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
+			   struct fid_av **fid_av, void *context);
+
+int xnet_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		     struct fid_domain **domain, void *context);
+
+static inline
+int xnet_create_subdomain(struct xnet_domain *domain, struct xnet_domain **subdomain)
```
Comment (aingerson) [inline prov/tcp/src/xnet.h]: I don't think we ever call create_subdomain anymore - I think we change it to just call fi_domain directly

File: prov/tcp/src/xnet_av.c [hunk 4]
```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2017-2022 Intel Corporation. All rights reserved.
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c]: Drop years

File: prov/tcp/src/xnet_av.c [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c]: Could be wrong but double check that this header is needed. I don't see an obvious reason for it

File: prov/tcp/src/xnet_av.c [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	ret = ofi_ip_av_insertsym(&av->util_av.av_fid, node, nodecnt,
+				  service, svccnt, fi_addr, flags, context);
+out:
+	ofi_genlock_unlock(&av->lock);
+
+	return ret;
+}
+
+int xnet_mplex_av_insertsvc(struct fid_av *av_fid, const char *node,
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c]: Static

File: prov/tcp/src/xnet_av.c [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+			goto out;
+	}
+	ret = ofi_ip_av_insertsvc(&av->util_av.av_fid, node, service,
+				  fi_addr, flags, context);
+out:
+	ofi_genlock_unlock(&av->lock);
+	return ret;
+}
+
+int xnet_mplex_av_lookup(struct fid_av *av_fid, fi_addr_t fi_addr,
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c]: static

File: prov/tcp/src/xnet_av.c [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+	return ofi_ip_av_lookup(&av->util_av.av_fid, fi_addr, addr, addrlen);
+}
+
+static const char *
+xnet_mplex_av_straddr(struct fid_av *av_fid, const void *addr, char *buf, size_t *len)
+{
+	return ofi_ip_av_straddr(av_fid, addr, buf, len);
+}
+
+int xnet_mplex_av_set(struct fid_av *av_fid, struct fi_av_set_attr *attr,
```
Comment (aingerson) [inline prov/tcp/src/xnet_av.c]: Also static

File: prov/tcp/src/xnet_domain.c [hunk 9]
```diff
@@ -128,6 +169,137 @@ xnet_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
...
+	struct fid_list_entry *item;
+	struct fid_mr *sub_mr_fid;
+	struct ofi_mr *mr;
+	int ret;
+
+	domain = container_of(fid, struct xnet_domain,
+			      util_domain.domain_fid.fid);
+	ret = ofi_mr_reg(fid, buf, len, access, offset, requested_key, flags,
+			 mr_fid, context);
+
```
Comment (aingerson) [inline prov/tcp/src/xnet_domain.c]: Drop this line

File: prov/tcp/src/xnet_domain.c [hunk 10]
```diff
@@ -128,6 +169,137 @@ xnet_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
...
+		}
+	}
+unlock:
+	ofi_genlock_unlock(&domain->subdomain_list_lock);
+out:
+	return ret;
+}
+
+static int
+xnet_mplex_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
```
Comment (aingerson) [inline prov/tcp/src/xnet_domain.c]: Can we squash this code (regattr/regv/reg) into one common function? Lots of duplication here

File: prov/tcp/src/xnet_domain.c [hunk 11]
```diff
@@ -167,9 +332,45 @@ xnet_query_atomic(struct fid_domain *domain, enum fi_datatype datatype,
...
+	return FI_SUCCESS;
+}
+
+static int xnet_mplex_domain_close(fid_t fid)
+{
+	struct xnet_domain *domain;
+	struct fid_list_entry *item;
+
+	domain = container_of(fid, struct xnet_domain, util_domain.domain_fid.fid);
+	while(!dlist_empty(&domain->subdomain_list)) {
```
Comment (aingerson) [inline prov/tcp/src/xnet_domain.c]: space after while

File: prov/tcp/src/xnet_domain.c [hunk 12]
```diff
@@ -226,12 +489,18 @@ int xnet_domain_open(struct fid_fabric *fabric_fid, struct fi_info *info,
...
+		return xnet_domain_multiplex_open(fabric_fid, info, domain_fid,
+						  context);
+
 	domain = calloc(1, sizeof(*domain));
 	if (!domain)
 		return -FI_ENOMEM;
 
 	ret = ofi_domain_init(fabric_fid, info, &domain->util_domain, context,
-			      OFI_LOCK_NONE);
+			      info->domain_attr->threading == FI_THREAD_SAFE ?
```
Comment (aingerson) [inline prov/tcp/src/xnet_domain.c]: I can't remember why we added this - was it a separate optimization or was it needed for these changes?

File: prov/tcp/src/xnet_progress.c [hunk 13]
```diff
@@ -908,9 +909,10 @@ static int xnet_handle_read_req(struct xnet_ep *ep)
 
 	resp->iov_cnt = 1 + resp->hdr.base_hdr.rma_iov_cnt;
 	resp->hdr.base_hdr.size = resp->iov[0].iov_len;
+	domain = container_of(ep->util_ep.domain, struct xnet_domain, util_domain);
 	for (i = 0; i < resp->hdr.base_hdr.rma_iov_cnt; i++) {
-		ret = ofi_mr_verify(&ep->util_ep.domain->mr_map, rma_iov[i].len,
-				    (uintptr_t *) &rma_iov[i].addr,
+		ret = ofi_mr_verify(&domain->util_domain.mr_map,
```
Comment (aingerson) [inline prov/tcp/src/xnet_progress.c]: I think this change was added before we moved everything to use the xnet_domain so I think we can take this out

File: prov/tcp/src/xnet_rdm.c [hunk 14]
```diff
@@ -695,13 +695,182 @@ static struct fi_ops_ep xnet_rdm_ep_ops = {
...
+	struct xnet_mplex_av *xnet_av;
+	struct util_av *util_av;
+	struct fid_av *av_fid;
+	struct fi_av_attr av_attr = {
+		.type = ep->domain->av_type,
+		.count = ep->av->av_entry_pool->entry_cnt,
+		.flags = 0,
+	};
+	size_t addr_size;
+	char addr[sizeof(struct sockaddr_in6)];
```
Comment (aingerson) [inline prov/tcp/src/xnet_rdm.c]: add a line after this

File: prov/tcp/src/xnet_rdm.c [hunk 15]
```diff
@@ -695,13 +695,182 @@ static struct fi_ops_ep xnet_rdm_ep_ops = {
...
+		if (!xnet_domain_multiplexed(&cntr->domain->domain_fid)) {
+			return container_of(&cntr->domain->domain_fid,
+					 struct xnet_domain, util_domain.domain_fid);
+		}
+	}
+
+	return NULL;
+}
+
+void set_subdomain(struct xnet_rdm *rdm, struct xnet_domain *domain,
```
Comment (aingerson) [inline prov/tcp/src/xnet_rdm.c]: static

File: prov/util/src/util_ep.c [hunk 16]
```diff
@@ -41,6 +41,8 @@ int ofi_ep_bind_cq(struct util_ep *ep, struct util_cq *cq, uint64_t flags)
 {
 	int ret;
 
+	assert(ep->domain == cq->domain);
```
Comment (aingerson) [inline prov/util/src/util_ep.c:44]: should we add the same assert to the counter bind path?

**Review by aingerson — COMMENTED** [review]
Sorry it took so long. Wanted to be thorough and also kept getting distracted... I think it looks good! Can't see any glaring issues

File: prov/tcp/src/xnet.h (hunk 2 above)
Comment (ooststep) [reply prov/tcp/src/xnet.h]: xnet_domain_open is the original default, we add the check to see if the domain is multiplexed there.  it already existed, just moved a few lines (see line 547 on the old version of this file)

File: prov/tcp/src/xnet_domain.c (hunk 12 above)
Comment (ooststep) [reply prov/tcp/src/xnet_domain.c]: I think this should actually be removed.   think it's left over from some earlier changes and likely has no real affect (since our new changes only apply to FI_THREAD_COMPLETION and this code is in the non-FI_THREAD_COMPLETION path, it should remain unchanged)

File: prov/tcp/src/xnet.h (hunk 2 above)
Comment (aingerson) [reply prov/tcp/src/xnet.h]: Oh I'm talking about lines 499-503 in the new file - it looks like you added it twice up here. But I don't think they need to move because I think they had to move to call in the new static inline for create_subdomains which is never called. So think this whole block can be dropped except the xnet_domain_multiplexed call. At least that's how I'm reading it

File: prov/tcp/src/xnet_av.c [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+						util_av.av_fid.fid);
+
+	ofi_genlock_lock(&av->lock);
+	dlist_foreach_container(&av->subav_list, struct fid_list_entry,	item, entry) {
+		subav_fid = container_of(item->fid, struct fid_av, fid);
+		ret = fi_av_insert(subav_fid, addr, count, fi_addr, flags, context);
+		if (ret < count)
+			goto out;
+	}
+	ret = ofi_ip_av_insert(&av->util_av.av_fid, addr, count, fi_addr, flags, context);
```
[the remark this answers is not in the document]
Comment (ooststep) [reply prov/tcp/src/xnet_av.c:83]: here, av->lock is the xnet_mplex_av:lock.  ip_av_insert_addr uses the util_av:lock

File: prov/tcp/src/xnet.h [hunk 18]
```diff
@@ -467,12 +467,43 @@ struct xnet_xfer_entry {
...
+	// progress engine and avoid having a single
+	// synchronization point among all eps.
+	struct fi_info		*subdomain_info;
+	struct ofi_genlock	subdomain_list_lock;
+	struct dlist_entry	subdomain_list;
 };
 
+int xnet_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		     struct fid_domain **domain, void *context);
+
```
Comment (aingerson) [inline prov/tcp/src/xnet.h]: I would move this back down since it's not needed anymore (unless you want to have it up here for organization purposes 🤷‍♀️ )

File: prov/util/src/util_mr_map.c [hunk 19]
```diff
@@ -52,7 +52,6 @@ dup_mr_attr(const struct fi_mr_attr *attr, uint64_t flags)
 	dup_attr->mr_iov = (struct iovec *) (dup_attr + 1);
 
 	/*
-	 * dup_mr_attr is only used insided ofi_mr_map_insert.
```
Comment (aingerson) [inline prov/util/src/util_mr_map.c]: sorry one last comment - this was changed from static and declared in the header but I don't see it called in this PR?

Comment (ooststep) [reply prov/util/src/util_mr_map.c]: probably leftover crud from a prior attempt. will revert
